src/Data/Yaml/Combinators.hs

```diff
@@ -186,6 +188,26 @@ valueConNames =
 fromComponent :: forall a . NS (ParserComponent a) (Code Value) -> Parser a
 fromComponent parser = Parser $ hexpand mempty parser
 
+-- Wrap a parser with a decorator. The decorator has access to the parsed value as well
+-- as the original and can inject its own processing logic.
+decorate :: forall a. Parser a -> (a -> Value -> Either ParseError a) -> Parser a
+decorate parser decorator = objectParser <> arrayParser <> stringParser <> numberParser <> boolParser
+  where
+    decorateComponent :: (x -> Value) -> NP I '[x] -> Either ParseError a
+    decorateComponent constructor (I x :* Nil) = case runParser parser orig of
+                                                    Left err  -> Left err
+                                                    Right val -> decorator val orig
+                                                    where
+                                                      orig = constructor x
+
+    -- There is a missing Null parser because it takes no arguments and breaks the pattern established by
+    -- these parsers.
+    objectParser = fromComponent $ Z                 $ ParserComponent $ Just $ decorateComponent Object
+    arrayParser  = fromComponent $ S . Z             $ ParserComponent $ Just $ decorateComponent Array
+    stringParser = fromComponent $ S . S . Z         $ ParserComponent $ Just $ decorateComponent String
+    numberParser = fromComponent $ S . S . S . Z     $ ParserComponent $ Just $ decorateComponent Number
+    boolParser   = fromComponent $ S . S . S . S . Z $ ParserComponent $ Just $ decorateComponent Bool
+
```

Comment:
Ok. So the new parser should only contain decorated components in the positions that the old one did. In retrospect I guess that seems obvious...